overlord/snapstate/backend/setup.go

```diff
@@ -30,15 +30,27 @@ import (
 )
 
 // SetupSnap does prepare and mount the snap for further processing.
-func (b Backend) SetupSnap(snapFilePath string, sideInfo *snap.SideInfo, meter progress.Meter) error {
+func (b Backend) SetupSnap(snapFilePath string, sideInfo *snap.SideInfo, meter progress.Meter) (err error) {
 	// This assumes that the snap was already verified or --dangerous was used.
 
-	s, snapf, err := OpenSnapFile(snapFilePath, sideInfo)
+	var s *snap.Info
+	var snapf snap.Container
+	s, snapf, err = OpenSnapFile(snapFilePath, sideInfo)
 	if err != nil {
 		return err
 	}
 	instdir := s.MountDir()
 
+	defer func() {
+		if err == nil {
+			return
+		}
+		// XXX: this will also remove the snap from /var/lib/snapd/snaps
+		if e := b.RemoveSnapFiles(s, s.Type, meter); e != nil {
```

Comment:
Not sure that we should do this. The opposite of setup is not to remove the whole snap. Does this tie correctly into the state (snap configuration and other things we keep)?

Reply:
FWIW, that's what we do on undoMountSnap.

Reply:
To clarify my previous comment, undoMountSnap calls backend.UndoSetupSnap, which does backend.RemoveSnapFiles(), so it does the same thing.

Reply:
@stolowski is correct, and the first argument passed to RemoveSnapFiles, despite being different beasts, return the same `MountFile()`, which is what' used to remove the snap.

@zyga is probably _also_ correct in that if a task didn't create the file it has no business removing it (that should be handled by the task that created it!), it's a separate issue that should be addressed separately -- and as @stolowski indirectly points out, needs addressing in more than what this PR is looking at.